addons/point_of_sale/static/src/app/services/pos_store.js

```diff
@@ -2663,6 +2674,302 @@ export class PosStore extends WithLazyGetterTrap {
         });
         await validation.validateOrder(false);
     }
+
+    handlePreparationHistory(srcPrep, destPrep, srcLine, destLine, qty) {
+        const srcKey = srcLine.preparationKey;
+        const destKey = destLine.preparationKey;
+        const srcQty = srcPrep[srcKey]?.quantity;
+
+        if (srcQty) {
+            if (srcQty <= qty) {
+                const newPrep = { ...srcPrep[srcKey], uuid: destLine.uuid };
+                destPrep[destKey] = newPrep;
+                delete srcPrep[srcKey];
+            } else {
+                srcPrep[srcKey].quantity = srcQty - qty;
+                destPrep[destKey] = { ...srcPrep[srcKey], uuid: destLine.uuid, quantity: qty };
+            }
+        }
+    }
+
+    get isSelectedLineCombo() {
+        const selectedOrderline = this.selectedOrder.getSelectedOrderline();
+        return selectedOrderline && selectedOrderline.isPartOfCombo();
+    }
+
+    getApplicableProductCombo(mode = "limited", productTmpl = null) {
+        const matchingCombos = [];
+        const productInOrder = this.selectedOrder.lines.reduce((acc, line) => {
+            if (line.isPartOfCombo()) {
+                return acc;
+            }
+            const pid = line.product_id.id;
+
+            if (!acc[pid]) {
+                acc[pid] = {
+                    lines: {},
+                    totalQty: 0,
+                };
+            }
+
+            acc[pid].lines[line.uuid] = line.qty;
+            acc[pid].totalQty += line.qty;
+            return acc;
+        }, {});
+        const combos = this.models["product.combo"].getAll();
+        const comboItems = combos.flatMap((combo) => combo.combo_item_ids);
+        const totalQtyAvailable = comboItems.reduce((acc, item) => {
+            const productId = item.product_id.id;
+            if (productInOrder[productId]) {
+                acc[item.combo_id.id] =
+                    (acc[item.combo_id.id] || 0) + productInOrder[productId].totalQty;
+            }
+            return acc;
+        }, {});
+
+        const productTmplsToCheck =
+            mode === "full" && productTmpl
+                ? [productTmpl]
+                : this.productCombos.sort(
+                      (a, b) =>
+                          a.product_tmpl_id.getPrice(this.selectedOrder.pricelist_id, 1) -
+                          b.product_tmpl_id.getPrice(this.selectedOrder.pricelist_id, 1)
+                  );
```

Comment:
Why comparing by prices ? If the goal is to find the most advantageous one, i don't think its necessary. In addition, getPrice is relatively complex to calculate.